### ISSUE TITLE
Consider not to suggest quoting HOME, UID, etc.

### ISSUE THREAD

**Issue by forthrin** [koalaman/shellcheck#2706]
#### For bugs
- Rule Id: SC2086 
- My shellcheck version: 0.9.0
- [X] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [X] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
touch /var/mail/$USER
```

#### Here's what shellcheck currently says:

`SC2086 Double quote to prevent globbing and word splitting.`

#### Here's what I wanted or expected to see:

Should this check be lenient on standard environment variables that won't have problematic characters in any sane environment? (`HOME UID LOGNAME RANDOM COLUMNS LINES`, etc.). `touch /var/mail/"$USER"` seems a bit redundant.

**Comment by brother** [comment]
```
brother ~$ echo $HOME
/home/brother
brother ~$ mkdir "/tmp/apa gnu"
brother ~$ unset HOME
brother /home/brother$ export HOME=/tmp/apa\ gnu/
brother /home/brother$ echo $HOME 
/tmp/apa gnu/
brother /home/brother$ cd $HOME
bash: cd: too many arguments
brother /home/brother$ LANG=C cd "$HOME"
```

**Comment by forthrin** [comment]
Which is exactly why I wrote "sane environment".

**Comment by brother** [comment]
And the point is that you can not dictate what that is.

**Comment by forthrin** [comment]
[Typiskt svenskt](https://museumofpsychology.files.wordpress.com/2019/03/man-cutting-the-branch-sitting-on-illustration-by-frits-ahlefeldt-1.jpg)

**Comment by Atrate** [comment]
If anything, it will get you into the good habit of always quoting variables, and in this case it has no downside.

**Comment by forthrin** [comment]
It has the downside of demanding quoting everything, when in fact only a vast minority of ENVs will ever have spaces, which will turn you against the feature in time and possibly disabling it altogether for peace of mind, with damning consequences.

Of the about hundred ENVs on this system, only a few have spaces (counting prompts as one). **Suggesting** thusly a pragmatic black/white lists that leaves the many obviously space-free ones alone, and enforces few the space-prone ones.



**LEGACY**
IFS=$' \t\n'
NAME='Major Dude'

**PROGRAMS**
ZIPOPT='-n .zip'
RUSTFLAGS='-A invalid_reference_casting'

**PROMPTS**
PS0='PS0 '
PS1='\W[\j] '
PS2='PS2$ '
PS3='Select: '
PS4='PS4$ '

**ARRAYS**
BASH_VERSINFO=([0]="3" [1]="2" ...)
PIPESTATUS=([0]="0" [1]="0" ...)

**Comment by forthrin** [comment]
@brother: Vad säger guldbaggen?

**Comment by wileyhy** [comment]
A lot of people use Shellcheck as a sort of tutoring program, for better or
worse. What you're suggesting is perhaps appropriate for advanced users,
but a newbie might find it more confusing than less confusing, in my
opinion.

Bash doesn't enforce parameter types, beyond basic attributes like
"readonly." It might be unorthodox, but...

~ $ UID=' WTH '
bash: UID: readonly variable
~ $ HOME=' WTH '
.../files/home $ echo $?
0
.../files/home $ declare -p HOME
declare -x HOME=" WTH "
.../files/home $

...repurposing environment variables for "off-label" uses can be valid code.

Wiley

**Comment by forthrin** [comment]
Linters are commonly integrated into editors, and go from useful to nuisance if they overdo their job. First thought of a white/black-list scheme, but actually the linter could simply just check your ENV. The following would seem pragmatic:

| **Case** | **Scope** | **Nature** | **Behavior** | **Warn**?       |
|----------|-----------|------------|----------------|-----------------|
| Upper    | Global    | Static     | Predictable    | If space in ENV |
| Lower    | Local     | Dynamic    | Unpredictable  | Always          |

**Comment by Atrate** [comment]
> Linters are commonly integrated into editors, and while like (Rubo)cops they are designed to protect and serve, they become nuisances and a threat to personal well-being (if not national security) once they extend their mandate beyond what's in the public interest.

Wow, what a lofty sentence talking about... *checks notes*, adding quotes around variables??

> First thought of a white/black-list scheme, but actually the linter could simply just check your ENV

Why though. This adds complexity, would be prone to errors and need future maintenance for literally no gain (possibly negative gain by teaching anti-patterns).

**Comment by forthrin** [comment]
Adding excess fluff quotes around 95% of global variables that will never need it, that is. But from the imminent scoffs of absolute correctness it seems safe to say this feature will never happen.

**Comment by wileyhy** [comment]
forthrin left a comment
...they become nuisances and a threat to personal well-being...


Oh, right, you mean like powdered fentanyl wrapped in a five dollar bill
and dropped someplace where playing children might find it "nuisance," or
stabbed 15 times,  dismembered and left unnoticed to rot for 7 days in a
full apartment complex "threat to personal well-being...."

You know, real life and death stuff, right?

**Comment by wileyhy** [comment]
"I said speak your mind, Jack, but __Jesus!__"

**Comment by forthrin** [comment]
Project maintainer can reopen the issue if they want to look into a solution in a constructive manner.

**Comment by wileyhy** [comment]
I had a really rough day, but I had no right to behave that way. I'm sorry.

**Comment by forthrin** [comment]
Ah, the slow burn of humility. 294 days to be exact. How nice to see the odd polar opposite of BOFH in the digital domain.

Obviously national security and friends was hyperbole, topped off with unnecessary insistence.

[SC2086](https://github.com/koalaman/shellcheck/issues?q=is%3Aissue+state%3Aopen+in:title+SC2086) is one of the very topmost reported error codes of all with 60/200 open issues, so no wonder the toilet roll facing debate.

We can only await the maintainer (ie. his LLM) to review them all and commit a super-patch that makes everyone one big happy family.

**Comment by wileyhy** [comment]
Um, no. That was just some hilation from very real day job, thanks. That
apology is for the ShellCheck crew.  :-|

Maybe some more reading would help you.

**Comment by forthrin** [comment]
... and we're back. But I'll take the good words intended for others any day.
